src/daemon/rpmostreed-daemon.cxx

```diff
@@ -726,21 +726,34 @@ char *
 rpmostreed_daemon_client_get_agent_id (RpmostreedDaemon *self, const char *client)
 {
   auto clientdata = static_cast<struct RpmOstreeClient *>(g_hash_table_lookup (self->bus_clients, client));
-  if (!clientdata || clientdata->id == NULL || g_str_equal (clientdata->id, "cli"))
+  struct RpmOstreeClient *clientdata_owned = NULL;
+  /* If the caller didn't register (e.g. Zincati doesn't today because it runs as a
+   * separate UID which systemd doesn't consider `active`), then let's gather the
+   * relevant client info now.
+   */
+  if (!clientdata)
+    clientdata = clientdata_owned = client_new (self, client, NULL);
+  if (clientdata->id == NULL || g_str_equal (clientdata->id, "cli"))
     return NULL;
```

Comment:
The logic here is confusing; we end up just returning what the caller passed in `client` right?

It looks like there's similar logic in `on_active_txn_changed()` which should apply here too right?

Hmm I think a root problem here is we are overlapping/conflating "client" (basically a process that has claimed a reference on the daemon to prevent auto-exiting) with "caller" (any dbus process talking to us).

What about something more like this?

```
diff --git a/src/daemon/rpmostreed-daemon.cxx b/src/daemon/rpmostreed-daemon.cxx
index e6df3cb2..04eb3dc1 100644
--- a/src/daemon/rpmostreed-daemon.cxx
+++ b/src/daemon/rpmostreed-daemon.cxx
@@ -250,7 +250,10 @@ on_active_txn_changed (GObject *object,
            * then let's gather the relevant client info now.
            */
           if (!clientdata)
-            clientdata = clientdata_owned = client_new (self, sender, NULL);
+            {
+              rpmostreed_daemon_add_client (self, sender, "cli");
+              clientdata = static_cast<struct RpmOstreeClient *>(g_hash_table_lookup (self->bus_clients, sender));
+            }
           g_autofree char *client_str = rpmostree_client_to_string (clientdata);
           g_autofree char *client_data_msg = NULL;
           if (clientdata->uid_valid)
```

But note the problem here is that without `RegisterClient` we lose the `zincati` string and we're actually back to figuring out it's `zincati.service` from the systemd unit.

Reply:
Alternatively/in addition: I would say that anything that is an update driver should have permission to invoke `RegisterClient`.  So we could change the [zincati polkit policy](https://github.com/coreos/zincati/blob/master/dist/polkit-1/rules.d/zincati.rules) to add `org.projectatomic.rpmostree1.client-management` - but then we'd also need to add something like an environment variable or CLI flag to force on registration attempt.

(A problem here is that the CLI is sort of crudely replicating the polkit rules to avoid doing something it thinks will fail)

Reply:
> we end up just returning what the caller passed in client right?

Not sure if I'm understanding correctly. But we basically want to take client (a DBus address), look up its PID, and get its systemd unit from that PID. So I guess we can skip a lot of the stuff that happens in `client_new()`, and then wouldn't need to do `rpmostree_client_free (clientdata_owned)`.
But you're suggestion is to actually add that DBus caller to the `g_hash_table` of "registered clients"?

Reply:
Ah ok, I see. This PR was originally intended to just change the `rpmostreed_daemon_client_get_sd_unit` function. You're right, in `rpmostreed_daemon_client_get_agent_id`, we do just end up returning whatever the caller passed in `client`.
My bad, I blindly copied the changes in `rpmostreed_daemon_client_get_sd_unit` to `rpmostreed_daemon_client_get_agent_id`. So it looks like we actually don't need to do anything in `rpmostreed_daemon_client_get_agent_id` for now, since we made `--register-driver` take in a string argument.